### PR TITLE
[SPARK-38001][SQL] Replace the error classes related to unsupported features by `UNSUPPORTED_FEATURE`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to re-use one error class `UNSUPPORTED_FEATURE` in the following Spark's exceptions:
- `QueryCompilationErrors.unsupportedIfNotExistsError` - when `IF NOT EXISTS` is not supported by `INSERT INTO`.
- `QueryExecutionErrors.aesModeUnsupportedError` - when an user specify unsupported AES mode and padding.
- `QueryExecutionErrors.literalTypeUnsupportedError` - impossible to create a literal from the input value (some Java class, for instance).
- `QueryExecutionErrors.transactionUnsupportedByJdbcServerError` - the target JDBC server does not support transaction.

And replace the following exceptions by `IllegalStateException` since they are internal, and should be not visible to users:
- `QueryExecutionErrors.simpleStringWithNodeIdUnsupportedError` - a sub-class of `Expression` or `Block` doesn't implements the method `simpleStringWithNodeId()`.
- `QueryExecutionErrors.dataTypeUnsupportedForWriterFuncError` - generating of a writer function for a struct field, array element, map key or map value doesn't support Catalyst's type.

Also, added new base test suite `QueryCompilationErrorsDSv2Suite` for testing DSv2 specific compilation error classes.

### Why are the changes needed?
Reducing the number of error classes should prevent from explode of `error-classes.json`. Also, using one error class for similar errors should improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
By running the affected test suites:
```
$ build/sbt "test:testOnly *SparkThrowableSuite"
$ build/sbt "test:testOnly *QueryExecutionErrorsSuite"
$ build/sbt "test:testOnly *DataSourceV2SQLSuite"
$ build/sbt "test:testOnly *DataSourceV2DataFrameSessionCatalogSuite"
$ build/sbt "test:testOnly *DataFramePivotSuite"
```
and the new test suite:
```
$ build/sbt "test:testOnly *QueryCompilationErrorsDSv2Suite"
```

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala [hunk 1]
```diff
@@ -254,7 +254,7 @@ object InterpretedUnsafeProjection {
         (_, _) => {}
 
       case _ =>
-        throw QueryExecutionErrors.dataTypeUnsupportedError(dt)
+        throw QueryExecutionErrors.dataTypeUnsupportedForWriterFuncError(dt)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala:257]: seems we can never reach here unless we have a bug. maybe we don't need error class here and can throw `IllegalStateException` directly?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedUnsafeProjection.scala:257]: How about to not handle the default case at all (I mean remove it). The compiler should output a warning if we forget to handle a type, and we.will get a match error exception in the case too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 2]
```diff
@@ -93,8 +93,8 @@ object QueryCompilationErrors {
 
   def unsupportedIfNotExistsError(tableName: String): Throwable = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:94]: shall we add a `QueryCompilationErrorsSuite` and test this error?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -132,8 +127,9 @@ object QueryExecutionErrors {
   }
 
   def simpleStringWithNodeIdUnsupportedError(nodeName: String): Throwable = {
-    new SparkUnsupportedOperationException(errorClass = "UNSUPPORTED_SIMPLE_STRING_WITH_NODE_ID",
-      messageParameters = Array(nodeName))
+    new SparkUnsupportedOperationException(
+      errorClass = "UNSUPPORTED_FEATURE",
+      messageParameters = Array(s"$nodeName does not implement simpleStringWithNodeId"))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:132]: this does not sound like a user-facing error. Can we double-check it?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:132]: Seems, it is not. At least, I haven't found a way how to trigger it from user level. I will replace the exception by `IllegalStateException` as for https://github.com/apache/spark/pull/35302#discussion_r792336423.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:94]: `InsertIntoSQLOnlyTests` has a test for the error already. BTW, the test from `InsertIntoSQLOnlyTests` runs as part of:

- V1WriteFallbackSessionCatalogSuite
- DataSourceV2SQLSuite
- DataSourceV2SQLSessionCatalogSuite
- DataSourceV2DataFrameSuite
- DataSourceV2DataFrameSessionCatalogSuite

@cloud-fan How about to create the `QueryCompilationErrorsSuiteBase` trait/abstract class and include it to those test suites?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:94]: I added the base test suites `QueryCompilationErrorsSuiteBase` + `SessionCatalogTestBase` and new test suites for compilation errors:

- QueryCompilationErrorsDSv2Suite
- QueryCompilationErrorsDSv2SessionCatalogSuite
- QueryCompilationErrorsV1WriteFallbackSuite

**Comment by MaxGekk** [comment]
@cloud-fan Could you look at this one more time, please.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 4]
```diff
@@ -112,4 +113,26 @@ class QueryExecutionErrorsSuite extends QueryTest with SharedSparkSession {
...
+    checkUnsupportedTypeInLiteral(("mike", 29, 1.0))
+
+    val e2 = intercept[SparkRuntimeException] {
+      trainingSales
+        .groupBy($"sales.year")
+        .pivot(struct(lower($"sales.course"), $"training"))
+        .agg(sum($"sales.earnings"))
+        .collect()
+    }
+    assert(e2.getMessage === "The feature is not supported: " +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:134]: TBH, this message is quite misleading, as the query does not create literals directly. We can improve it later.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
jenkins merge

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:134]: I opened the ticket SPARK-38097 to improve the error.

**Pull request closed without merging**